tests/framework/e2e/cluster.go

```diff
@@ -253,173 +271,173 @@ func (cfg *EtcdProcessClusterConfig) PeerScheme() string {
 	return setupScheme(cfg.BasePeerScheme, cfg.IsPeerTLS)
 }
 
-func (cfg *EtcdProcessClusterConfig) EtcdServerProcessConfigs(tb testing.TB) []*EtcdServerProcessConfig {
-	lg := zaptest.NewLogger(tb)
+func (cfg *EtcdProcessClusterConfig) EtcdAllServerProcessConfigs(tb testing.TB) []*EtcdServerProcessConfig {
+	etcdCfgs := make([]*EtcdServerProcessConfig, cfg.ClusterSize)
+	initialCluster := make([]string, cfg.ClusterSize)
 
-	if cfg.BasePort == 0 {
-		cfg.BasePort = EtcdProcessBasePort
-	}
-	if cfg.ExecPath == "" {
-		cfg.ExecPath = BinPath
+	for i := 0; i < cfg.ClusterSize; i++ {
+		etcdCfgs[i] = cfg.EtcdServerProcessConfig(tb, i)
+		initialCluster[i] = fmt.Sprintf("%s=%s", etcdCfgs[i].Name, etcdCfgs[i].Purl.String())
 	}
-	if cfg.SnapshotCount == 0 {
-		cfg.SnapshotCount = etcdserver.DefaultSnapshotCount
+
+	initialClusterArgs := []string{"--initial-cluster", strings.Join(initialCluster, ",")}
+	for i := range etcdCfgs {
+		etcdCfgs[i].SetInitialCluster(initialCluster, "new")
+		etcdCfgs[i].InitialCluster = strings.Join(initialCluster, ",")
+		etcdCfgs[i].Args = append(etcdCfgs[i].Args, initialClusterArgs...)
 	}
 
-	etcdCfgs := make([]*EtcdServerProcessConfig, cfg.ClusterSize)
-	initialCluster := make([]string, cfg.ClusterSize)
-	for i := 0; i < cfg.ClusterSize; i++ {
-		var curls []string
-		var curl string
-		port := cfg.BasePort + 5*i
-		clientPort := port
-		peerPort := port + 1
-		peer2Port := port + 3
-		clientHttpPort := port + 4
-
-		if cfg.ClientTLS == ClientTLSAndNonTLS {
-			curl = clientURL(cfg.ClientScheme(), clientPort, ClientNonTLS)
-			curls = []string{curl, clientURL(cfg.ClientScheme(), clientPort, ClientTLS)}
-		} else {
-			curl = clientURL(cfg.ClientScheme(), clientPort, cfg.ClientTLS)
-			curls = []string{curl}
-		}
+	return etcdCfgs
+}
 
-		purl := url.URL{Scheme: cfg.PeerScheme(), Host: fmt.Sprintf("localhost:%d", peerPort)}
-		peerAdvertiseUrl := url.URL{Scheme: cfg.PeerScheme(), Host: fmt.Sprintf("localhost:%d", peerPort)}
-		var proxyCfg *proxy.ServerConfig
-		if cfg.PeerProxy {
-			if !cfg.IsPeerTLS {
-				panic("Can't use peer proxy without peer TLS as it can result in malformed packets")
-			}
-			peerAdvertiseUrl.Host = fmt.Sprintf("localhost:%d", peer2Port)
-			proxyCfg = &proxy.ServerConfig{
-				Logger: zap.NewNop(),
-				To:     purl,
-				From:   peerAdvertiseUrl,
-			}
-		}
+func (cfg *EtcdServerProcessConfig) SetInitialCluster(nodes []string, initialClusterState string) {
```

Comment:
The implementation seems to be different from the main branch. There is no `SetInitialCluster`. Instead it's `SetInitialOrDiscovery`.

Also the implementation of `EtcdServerProcessConfig` is also a little different from the main branch.

Reply:
My understanding is that the PR #14589 is to extract the part originally in the for-loop as a new method `EtcdServerProcessConfig `. Then, using this method, we can create a config and dynamically add new members to the cluster. As for the implementation itself, it needs to backport some other PRs to align.

`SetInitialOrDiscovery` is also the same reason, `discovery` was introduced by https://github.com/etcd-io/etcd/pull/13621/commits/a53074542bdd62e139fc2c268e17c77caf16ef45, so I removed discovery from the method.

I can pause this PR first and align the content of `EtcdServerProcessConfig` and  `discovery`. What do you think?

Reply:
> I can pause this PR first

OK



> align the content of `EtcdServerProcessConfig` and `discovery`.

Not fully understood your point. but happy to have more discuss under your new PR.